### PR TITLE
Proposal for improving language support in 2019

### PR DESCRIPTION
**Pull request by chrismwendt**
Based on suggestions from our meeting today, I:

- Tied this proposal back to the Q1 2019 goal
- Added a note about how UI/UX matters
- Simplified the test plan, made it more concrete
- Added items to the checklist with (highly) tentative due dates

### CODE REVIEW

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+Put effort towards one language at a time in rough order of prospective deals for Sourcegraph:
+
+- ✅ JavaScript/TypeScript using https://github.com/sourcegraph/lang-typescript
+- 📝 Java using https://github.com/beyang/eclipse.jdt.ls/tree/wip
+- 📝 Python using https://github.com/sourcegraph/sourcegraph-python
+- ✅ Go using https://github.com/sourcegraph/lang-go
+- 📝 Swift using https://github.com/apple/sourcekit-lsp
+
+For each language, build a Sourcegraph extension that communicates with the language server, aiming to check boxes in the test plan.
```
Comment (sqs) [inline doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]:
```suggestion
For each language, patch the language server as needed and build a Sourcegraph extension that communicates with it and satisfies the test plan below.
```

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+### Test plan
+
+Manually open each link to verify that the language server is working. In rough order of increasing difficulty:
```
Comment (sqs) [inline doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: Clarify how this test plan works. When developing a language's support, the author should copy this test plan and customize it for their language. You @chrismwendt 	have final say on the repositories and links they get to use.

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+- [ ] Link to ANY working {hover,definition,references}
+- [ ] Link to 3 working {hover,definnition,references}, each in a different popular repo (e.g. for Go https://github.com/search?q=stars%3A%3E0+language%3Ago&type=Repositories)
```
Comment (sqs) [inline doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]:
```suggestion
- [ ] Link to 3 working {hover,definition,references}, each in a different popular repo (e.g. for Go https://github.com/search?q=stars%3A%3E0+language%3Ago&type=Repositories)
```

Comment (sqs) [reply doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: Also, "repository" not "repo" (https://github.com/sourcegraph/about/blob/master/STYLEGUIDE.md#specific-terms-in-prose)

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+- [ ] Initialization time on popular repos is <10s
+- [ ] After initialization on popular repos, most response times are <3s
+- [ ] Monorepo support (inintialization time and response time scale sublinearly with repo size)
+- [ ] Cross-repo code intelligence on popular repos
+
+### Release plan
+
+For each language, publish a Sourcegraph extension with usage instructions in their READMEs and then iterate and address feedback until customers actually use the language extension on a regular basis.
```
Comment (sqs) [inline doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: Break this into a numbered list for readability

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+### Test plan
+
+Manually open each link to verify that the language server is working. In rough order of increasing difficulty:
+
+- [ ] Link to ANY working {hover,definition,references}
+- [ ] Link to 3 working {hover,definnition,references}, each in a different popular repo (e.g. for Go https://github.com/search?q=stars%3A%3E0+language%3Ago&type=Repositories)
+- [ ] Initialization time on popular repos is <10s
+- [ ] After initialization on popular repos, most response times are <3s
+- [ ] Monorepo support (inintialization time and response time scale sublinearly with repo size)
```
Comment (sqs) [inline doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]:
```suggestion
- [ ] Monorepo support (initialization time and response time scale sublinearly with repo size)
```

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+- [ ] Cross-repo code intelligence on popular repos
```
Comment (sqs) [inline doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: Call this out as not yet fully specified and that you @chrismwendt 	will deal with each language on a case-by-case basis to determine the requirements for cross-repo code intel (possibly in consultation with @felixfbecker).

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+## Checklist
+
+- [ ] By Dec 21, publish a modified version of this proposal that communicates our high level plan and states what Sourcegraph wants in terms of language support. Briefly summarize and link to the master plan. Mention that we'll help with sponsorship, technical advice, getting feedback, etc.
+- [ ] By Dec 21, update langserver.org and our careers repo to link to this blog post and have @sqs to tweet this to Future of Coding
+- [ ] By Dec 21, after doing an initial triage, publish one blog post per language similar to the above blog post but with more details about the particular language.
+- [ ] By Jan 2, add an Implementation section to the README of lang-{go,typescript,python}
+- [ ] By Jan 2, prepare a doc outlining how to build language extensions building off of https://github.com/sourcegraph/sourcegraph/pull/628 but avoid recommending certain approaches (leave it up to the language server author), link to the Implementation sections above
+- [ ] By Jan 4, basic Python support is deployed to Sourcegraph.com (@sqs has already implemented most of this)
```
Comment (sqs) [inline doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: I will prepare Python to be handed off to you tomorrow (Fri) night

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+Now that we have removed LSP proxy and added Sourcegraph extensions, the landscape has changed:
+
+- With Sourcegraph extensions, it’s easier to understand how code intelligence on Sourcegraph works, which makes it easier to build on and all you need to understand is the Sourcegraph extension API (no need to understand xfiles/xcontents, xcache, lsp-adapter, lsp-proxy, etc.) in order to add language support
+- There’s a new Swift language server https://github.com/apple/sourcekit-lsp
+- There’s a new Python language server https://github.com/Microsoft/python-language-server
+- We’ve learned that it’s practical (technically) to add zip archive fetching and WebSocket support directly to language servers (implemented for Go, TypeScript, and Python), and zip archives of the whole repo are usually fast enough for our purposes (whether or not these will be upstreamed is mentioned under risks below)
+- We’ve learned that a poorly implemented language server is worse than no language server at all (forever loading hover tooltip spinner wastes time, empty or trivial hover tooltip is distracting, jump-to-definition not working is frustrating, and all of this reflected poorly on Sourcegraph)
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: The point I was making here is that there's a threshold level of quality below which a language server is actually detrimental to productivity, and many of the experimental language servers did not meet that threshold.

However, your point still stands: those experimental language servers probably still wouldn't have improved even if they did meet that threshold.

I'm committing your suggestion.

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+- Focus on quality over quantity (already tried quantity in 2018 with lsp-adapter and cold emails, and this stagnated)
+- Pay contractors if it saves us time and effort (they’re experts in the respective language and will likely be able to hit the ground running)
+- UI/UX ergonomics matter: suppress non-actionable errors and indicate when some analysis is taking a long time
+
+## Proposal
+
+Put effort towards one language at a time in rough order of prospective deals for Sourcegraph:
+
+- ✅ JavaScript/TypeScript using https://github.com/sourcegraph/lang-typescript
+- 📝 Java using https://github.com/beyang/eclipse.jdt.ls/tree/wip
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: https://github.com/sourcegraph/sourcegraph/pull/1503#issuecomment-448806852 Sounds like he's going to perf test eclipse.jdt.ls before reopening the discussion.

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+These principles will guide future improvements:
+
+- Prioritize languages by a combination of popularity and ease of analysis (statically-typed languages, ones that already have solid language servers, ones that customers are asking for, etc.)
+- Focus on quality over quantity (already tried quantity in 2018 with lsp-adapter and cold emails, and this stagnated)
+- Pay contractors if it saves us time and effort (they’re experts in the respective language and will likely be able to hit the ground running)
+- UI/UX ergonomics matter: suppress non-actionable errors and indicate when some analysis is taking a long time
+
+## Proposal
+
+Put effort towards one language at a time in rough order of prospective deals for Sourcegraph:
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: I consolidated the definition of priority to 1 place (in the principles section above) and changed it to include both of your suggestions: `a combination of popularity, ease of analysis, and Sourcegraph customer needs`.

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+For each language, build a Sourcegraph extension that communicates with the language server, aiming to check boxes in the test plan.
+
+### Test plan
+
+Manually open each link to verify that the language server is working. In rough order of increasing difficulty:
+
+- [ ] Link to ANY working {hover,definition,references}
+- [ ] Link to 3 working {hover,definnition,references}, each in a different popular repo (e.g. for Go https://github.com/search?q=stars%3A%3E0+language%3Ago&type=Repositories)
+- [ ] Initialization time on popular repos is <10s
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: Totally agree, I plan to pick ~3 non-trivial repositories (most likely libraries or command-line tools) from the front page of the GitHub repositories page sorted by stars. My goal is to pick repositories that are representative of the set of repositories that Sourcegraph users will likely be navigating (e.g. avoid picking trivial repositories with no dependencies or only picking repositories that all use the same build configuration).

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+
+For each language, publish a Sourcegraph extension with usage instructions in their READMEs and then iterate and address feedback until customers actually use the language extension on a regular basis.
+
+## Rationale
+
+[A discussion of alternate approaches and the trade offs, advantages, and disadvantages (including risks and uncertainties) of the specified approach.]
+
+## Checklist
+
+- [ ] By Dec 21, publish a modified version of this proposal that communicates our high level plan and states what Sourcegraph wants in terms of language support. Briefly summarize and link to the master plan. Mention that we'll help with sponsorship, technical advice, getting feedback, etc.
```
[the remark this answers is not in the document]
Comment (chrismwendt) [reply doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: Yes, this line refers to the blog post in https://github.com/sourcegraph/about/pull/69 I'll clarify this.

File: doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md (hunk 7 above)
Comment (chrismwendt) [reply doc/dev/proposals/2018-12-20-improving-language-support-in-2019.md]: Hovers/defs/refs already work. It sounds like there are bugs to iron out. It also needs to get deployed (i.e. added to the infrastructure repo). See the latest status update https://sourcegraph.slack.com/archives/C89KCDK5J/p1542193364089200

I will have a better estimate once I've tried it out more extensively and looked at the code a bit more https://github.com/Microsoft/python-language-server/compare/master...sourcegraph:wip2

**Pull request merged**